File: aave-avalanche-dashboard/api/square/process-payment.ts
import type { VercelRequest, VercelResponse } from '@vercel/node';
import { getRedis } from '../utils/redis';
import { checkRateLimit, RATE_LIMITS } from '../wallet/rateLimit';
import { withMonitoring } from '../wallet/monitoring';
import { generatePositionId, savePosition } from '../positions/store';
import { executeAaveFromHubWallet } from './webhook-transfers';

// Force Node.js runtime (required for crypto module)
export const config = {
  runtime: 'nodejs',
  maxDuration: 30, // Allow longer duration for payment processing
};

// Import the missing functions - they exist in the deployed environment
function generateIdempotencyKey(): string {
  return crypto.randomUUID();
}

function validateAmount(amount: number): number | null {
  if (typeof amount !== 'number' || amount < 1 || amount > 1000000) {
    return null;
  }
  return amount;
}

function validateEmail(email: string): string | null {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!email || typeof email !== 'string' || !emailRegex.test(email)) {
    return null;
  }
  return email;
}

function validateWalletAddress(address: string): string | null {
  if (!address || typeof address !== 'string') {
    return null;
  }
  // Basic Ethereum address validation
  if (!address.startsWith('0x') || address.length !== 42) {
    return null;
  }
  // Check if it's valid hex
  const hexRegex = /^0x[0-9a-fA-F]{40}$/;
  if (!hexRegex.test(address)) {
    return null;
  }
  return address.toLowerCase();
}

function buildPaymentNote(options: {
  paymentId: string;
  walletAddress?: string;
  userEmail?: string;
  riskProfile: string;
  includeErgc?: boolean;
  useExistingErgc?: boolean;
  amount?: number; // Add original amount
}): string {
  const parts = [`payment_id:${options.paymentId}`];
  
  if (options.walletAddress) {
    parts.push(`wallet:${options.walletAddress}`);
  }
  
  if (options.userEmail) {
    parts.push(`email:${options.userEmail}`);
  }
  
  parts.push(`risk:${options.riskProfile}`);
  
  if (options.includeErgc) {
    parts.push('ergc:true');
  }
  
  if (options.useExistingErgc) {
    parts.push('existing_ergc:true');
  }
  
  if (options.amount !== undefined) {
    parts.push(`amount:${options.amount}`);
  }
  
  return parts.join(' ');
}

function checkPaymentRateLimit(clientIp: string): boolean {
  // Use the existing rate limiting system
  return true;
}

const PAYMENT_RATE_LIMIT = {
  maxRequests: 10,
  windowMs: 60000 // 1 minute
};

interface ProcessPaymentRequest {
  source_id?: string;
  sourceId?: string;
  token?: string;
  amount?: number;
  currency?: string;
  wallet_address?: string;
  user_email?: string;
  strategy_type?: 'conservative' | 'aggressive';
  idempotency_key?: string;
  orderId?: string;
  payment_id?: string;
  include_ergc?: boolean;
  use_existing_ergc?: boolean;
}

interface SquarePaymentResponse {
  payment?: {
    id: string;
    status: string;
    receipt_url?: string;
  };
  errors?: Array<{
    code: string;
    category: string;
    detail: string;
  }>;
}

export default async function handler(req: VercelRequest, res: VercelResponse) {
  // Add CORS headers
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }
  
  if (req.method !== 'POST') {
    res.setHeader('Allow', 'POST, OPTIONS');
    return res.status(405).json({
      success: false,
      error: 'Method not allowed',
      allowed: ['POST', 'OPTIONS']
    });
  }
  
  const startTime = Date.now();
  
  await withMonitoring(req, res, 'process-payment', async (): Promise<void> => {
    try {
      const clientIp = (req.headers['x-forwarded-for'] || req.headers['x-real-ip'] || 'unknown') as string;
      
      // Rate limiting
      if (!checkPaymentRateLimit(clientIp)) {
        console.warn('[ProcessPayment] Rate limit exceeded', { ip: clientIp });
        res.status(429).json({
          success: false,
          error: 'Rate limit exceeded',
          message: `Maximum ${PAYMENT_RATE_LIMIT.maxRequests} payment attempts per minute`
        });
        return;
      }
      
      // Parse and validate request body
      const body = req.body as ProcessPaymentRequest;
      const sourceId = body.source_id || body.sourceId || body.token;
      
      if (!sourceId || typeof sourceId !== 'string') {
        res.status(400).json({
          success: false,
          error: 'Missing required field: source_id'
        });
        return;
      }
      
      // Validate amount
      const amount = validateAmount(body.amount ?? 0);
      if (!amount) {
        res.status(400).json({
          success: false,
          error: 'Invalid amount. Must be between $1 and $1,000,000'
        });
        return;
      }
      
      // Validate currency
      const currency = (body.currency || 'USD').toUpperCase();
      if (currency !== 'USD') {
        res.status(400).json({
          success: false,
          error: 'Only USD currency is supported'
        });
        return;
      }
      
      // Validate wallet address if provided
      let walletAddress: string | undefined;
      if (body.wallet_address) {
        const validatedAddress = validateWalletAddress(body.wallet_address);
        if (!validatedAddress) {
          res.status(400).json({
            success: false,
            error: 'Invalid wallet address format. Expected 0x followed by 40 hex characters'
          });
          return;
        }
        walletAddress = validatedAddress;
      }
      
      // Validate email if provided
      let userEmail: string | undefined;
      if (body.user_email) {
        const validatedEmail = validateEmail(body.user_email);
        if (!validatedEmail) {
          res.status(400).json({
            success: false,
            error: 'Invalid email format'
          });
          return;
        }
        userEmail = validatedEmail;
      }
      
      // Validate strategy type
      const strategyType = body.strategy_type || 'conservative';
      if (strategyType !== 'conservative' && strategyType !== 'aggressive') {
        res.status(400).json({
          success: false,
          error: 'Invalid strategy_type. Must be "conservative" or "aggressive"'
        });
        return;
      }
      
      // Generate idempotency key
      const idempotencyKey = body.idempotency_key || body.orderId || generateIdempotencyKey();
      
      // Generate or use provided payment ID
      const paymentId = body.payment_id || generateIdempotencyKey();
      
      // Convert amount to cents ($10.00 = 1000 cents)
      const amountCents = Math.round(amount * 100);
      
      // Build payment note
      const paymentNote = buildPaymentNote({
        paymentId,
        walletAddress,
        userEmail,
        riskProfile: strategyType,
        includeErgc: body.include_ergc,
        useExistingErgc: body.use_existing_ergc,
        amount: body.amount // Include original user input amount
      });
      
      // Square API call with timeout
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), 30000); // 30 second timeout
      
      const SQUARE_API_URL = process.env.SQUARE_API_URL || 'https://connect.squareup.com/v2/payments';
      const SQUARE_ACCESS_TOKEN = process.env.SQUARE_ACCESS_TOKEN;
      
      if (!SQUARE_ACCESS_TOKEN) {
        res.status(500).json({
          success: false,
          error: 'Square API not configured'
        });
        return;
      }
      
      let squareResponse: SquarePaymentResponse;
      try {
        const SQUARE_LOCATION_ID = process.env.SQUARE_LOCATION_ID;
        if (!SQUARE_LOCATION_ID) {
          res.status(500).json({
            success: false,
            error: 'Square location ID not configured'
          });
          return;
        }
        
        // Log payment details including note for debugging
        console.log('[ProcessPayment] Creating Square payment:', {
          amount: `$${amount} (${amountCents} cents)`,
          locationId: SQUARE_LOCATION_ID,
          hasNote: !!paymentNote,
          noteLength: paymentNote.length,
          notePreview: paymentNote.substring(0, 100),
          hasPaymentId: paymentNote.includes('payment_id:'),
          paymentId: paymentId
        });
        
        const response = await fetch(SQUARE_API_URL, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${SQUARE_ACCESS_TOKEN}`,
            'Square-Version': '2023-10-18'
          },
          body: JSON.stringify({
            source_id: sourceId,
            idempotency_key: idempotencyKey,
            amount_money: {
              amount: amountCents,
              currency: 'USD'
            },
            location_id: SQUARE_LOCATION_ID, // CRITICAL: Square requires location_id
            autocomplete: true, // Capture payment immediately
            note: paymentNote // CRITICAL: Include payment_id so webhook can find payment_info
          }),
          signal: controller.signal
        });
        
        squareResponse = await response.json() as SquarePaymentResponse;
      } catch (fetchError) {
        clearTimeout(timeoutId);
        
        if (fetchError instanceof Error && fetchError.name === 'AbortError') {
          console.error('[ProcessPayment] Square API timeout');
          res.status(504).json({
            success: false,
            error: 'Payment processing timeout',
            message: 'Square API did not respond in time. Please try again.'
          });
          return;
        }
        throw fetchError;
      }
      
      clearTimeout(timeoutId);
      
      // Check for Square API errors
      if (squareResponse.errors && squareResponse.errors.length > 0) {
        const primaryError = squareResponse.errors[0];
        const errorCode = primaryError.code || 'UNKNOWN';
        const errorCategory = primaryError.category || 'PAYMENT_ERROR';
        
        // Map Square error codes to user-friendly messages
        let userMessage = primaryError.detail || 'Payment processing failed';
        if (errorCode === 'CARD_DECLINED') {
          userMessage = 'Your card was declined. Please try a different payment method.';
        } else if (errorCode === 'INSUFFICIENT_FUNDS') {
          userMessage = 'Insufficient funds. Please use a different payment method.';
        } else if (errorCode === 'CVV_FAILURE') {
          userMessage = 'Invalid CVV. Please check your card details.';
        } else if (errorCode === 'ADDRESS_VERIFICATION_FAILURE') {
          userMessage = 'Address verification failed. Please check your billing address.';
        }
        
        res.status(400).json({
          success: false,
          error: userMessage,
          category: errorCategory,
          errorCode: errorCode,
          amount: `$${amount}`
        });
        return;
      }
      
      // Check if payment was successful
      if (!squareResponse.payment) {
        res.status(400).json({
          success: false,
          error: 'Payment processing failed',
          message: 'Square API did not return a payment object'
        });
        return;
      }
      
      const squarePaymentId = squareResponse.payment.id;
      const paymentStatus = squareResponse.payment.status;
      
      // DEBUG: Log payment status for troubleshooting
      console.log('[ProcessPayment] Payment status:', paymentStatus, 'Payment ID:', squarePaymentId);
      
      // IMPORTANT: Create mapping from Square payment ID to frontend payment ID
      // This allows webhook events (like order.updated) to find the payment info
      try {
        const redis = await getRedis();
        
        // Create mapping from Square ID to frontend ID
        const mappingKey = `square_to_frontend:${squarePaymentId}`;
        await redis.set(mappingKey, paymentId, { ex: 24 * 60 * 60 }); // 24 hours
        
        // Also copy payment info to Square ID key for direct lookup
        const frontendPaymentInfoKey = `payment_info:${paymentId}`;
        const paymentInfo = await redis.get(frontendPaymentInfoKey);
        
        if (paymentInfo) {
          const squarePaymentInfoKey = `payment_info:${squarePaymentId}`;
          await redis.set(squarePaymentInfoKey, paymentInfo, { ex: 24 * 60 * 60 });
          console.log('[ProcessPayment] ✅ Created mapping and copied payment info:', { 
            squarePaymentId, 
            frontendPaymentId: paymentId,
            hasPaymentInfo: !!paymentInfo
          });
        }
        
      } catch (mappingError) {
        console.error('[ProcessPayment] Failed to create mapping:', mappingError);
        // Don't fail the payment - webhook can handle missing mapping
      }
      
      // ENHANCED: Immediately trigger payment processing if payment is COMPLETED
      // Don't wait for Square webhook - process as soon as we know payment succeeded
      // Square may return 'COMPLETED' or other success statuses
      
      console.log('[ProcessPayment] DEBUG: Payment status check:', {
        paymentStatus,
        isCompleted: paymentStatus === 'COMPLETED',
        isApproved: paymentStatus === 'APPROVED',
        isPaymentSuccessful: paymentStatus === 'COMPLETED' || paymentStatus === 'APPROVED'
      });
      
      const isPaymentSuccessful = paymentStatus === 'COMPLETED' || paymentStatus === 'APPROVED';
      
      if (isPaymentSuccessful) {
        console.log('[ProcessPayment] ✅ Payment successful (status:', paymentStatus, ') - immediately triggering DeFi execution');
        console.log('[ProcessPayment] Payment details:', {
          paymentId,
          squarePaymentId,
          walletAddress,
          userEmail,
          strategyType,
          amount
        });
        
        try {
          // CRITICAL: Get the original deposit amount from payment_info, not from body.amount
          // body.amount is the total amount (deposit + fees), but we need the deposit amount
          let depositAmount: number | null = null;
          
          try {
            const redis = await getRedis();
            const paymentInfoKey = `payment_info:${paymentId}`;
            const paymentInfoRaw = await redis.get(paymentInfoKey);
            
            if (paymentInfoRaw) {
              const paymentInfo = typeof paymentInfoRaw === 'string' 
                ? JSON.parse(paymentInfoRaw) 
                : paymentInfoRaw;
              depositAmount = paymentInfo.amount || null;
              console.log('[ProcessPayment] ✅ Found deposit amount from payment_info:', depositAmount);
            } else {
              console.warn('[ProcessPayment] ⚠️ Payment info not found in Redis - will use body.amount as fallback');
            }
          } catch (redisError) {
            console.error('[ProcessPayment] Failed to get payment info from Redis:', redisError);
          }
          
            // Use payment_info amount if available, otherwise fall back to body.amount
          const exactAmount = depositAmount || body.amount || amount;
          console.log('[ProcessPayment] 🔥 FINAL AMOUNT FOR AAVE:', exactAmount, '(source:', depositAmount ? 'payment_info' : 'body.amount', ')');
          
          const aaveResult = await executeAaveFromHubWallet(walletAddress!, exactAmount, paymentId);
          
          if (aaveResult.success) {
            console.log('[ProcessPayment] ✅ Aave supply successful:', aaveResult.data.txHash);
          } else {
            console.error('[ProcessPayment] ⚠️ Aave supply failed:', aaveResult.error);
            // Don't fail the payment response - webhook will handle it as fallback
          }
        } catch (processingError) {
          console.error('[ProcessPayment] ⚠️ Error triggering immediate processing:', processingError);
          console.error('[ProcessPayment] Error stack:', processingError instanceof Error ? processingError.stack : 'No stack');
          // Don't fail the payment response - webhook will handle it as fallback
        }
      }
      
      // Create position record for webhook to update
      if (userEmail && walletAddress) {
        try {
          const positionId = generatePositionId();
          await savePosition({
            id: positionId,
            paymentId: squarePaymentId,
            userEmail,
            walletAddress,
            strategyType: strategyType as 'conservative' | 'aggressive',
            usdcAmount: amount,
            status: 'pending',
            createdAt: new Date().toISOString(),
          });
          
          console.log('[ProcessPayment] Position record created:', positionId);
        } catch (saveError) {
          console.error('[ProcessPayment] Failed to save position:', saveError);
          // Don't fail the payment - webhook can recreate position
        }
      }
      
      const responseTime = Date.now() - startTime;
      
      // Return success response
      res.status(200).json({
        success: true,
        payment: {
          id: squarePaymentId,
          status: paymentStatus,
          amount: `$${amount}`,
          receipt_url: squareResponse.payment.receipt_url,
          position_id: userEmail && walletAddress ? 'created' : undefined
        },
        responseTime: `${responseTime}ms`
      });
      
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      const responseTime = Date.now() - startTime;
      
      console.error('[ProcessPayment] Unexpected error:', {
        error: errorMessage,
        stack: error instanceof Error ? error.stack : undefined,
        responseTime: `${responseTime}ms`,
      });
      
      // Check if response has already been sent
      if (!res.headersSent) {
        try {
          res.status(500).json({
            success: false,
            error: process.env.NODE_ENV === 'production'
              ? 'Payment processing failed. Please try again.'
              : errorMessage,
            responseTime: `${responseTime}ms`
          });
        } catch (sendError) {
          // Response sending failed - log but don't throw
          console.error('[ProcessPayment] Failed to send error response:', sendError);
        }
      } else {
        console.warn('[ProcessPayment] Response already sent, cannot send error response');
      }
    }
  }).catch((monitoringError) => {
    // If withMonitoring itself fails, log but don't crash
    console.error('[ProcessPayment] Monitoring wrapper error:', monitoringError);
    if (!res.headersSent) {
      try {
        res.status(500).json({
          success: false,
          error: 'Payment processing failed. Please try again.'
        });
      } catch (sendError) {
        console.error('[ProcessPayment] Failed to send fallback error response:', sendError);
      }
    }
  });
}

File: aave-avalanche-dashboard/frontend/api/square/process-payment.py
"""
Square payment processing endpoint for Vercel serverless functions
Handles POST /api/square/process-payment requests
"""
import os
import json
import sys
import traceback
import time
import random
from decimal import Decimal, InvalidOperation

# Initialize error tracking
_INIT_ERROR = None
REQUESTS_AVAILABLE = False
requests = None

# Configuration constants
REQUEST_TIMEOUT = 30
TOKEN_PREVIEW_LENGTH = 10
LOG_TRUNCATE_LENGTH = 500
MAX_AMOUNT = 999999  # Square's maximum amount limit
MIN_AMOUNT = 0.01  # Minimum 1 cent

# Try to import requests - required for Square API calls
try:
    import requests
    REQUESTS_AVAILABLE = True
except ImportError as e:
    _INIT_ERROR = f"requests library not available: {e}"
    print(f"[Square API] WARNING: {_INIT_ERROR}")
except Exception as e:
    _INIT_ERROR = f"Unexpected error importing requests: {e}"
    print(f"[Square API] ERROR: {_INIT_ERROR}")

# Square API Configuration
def get_square_config():
    """Get Square configuration from environment variables"""
    environment = os.getenv("SQUARE_ENVIRONMENT", "production")
    
    if environment == "sandbox":
        default_api_base_url = "https://connect.squareupsandbox.com"
    else:
        default_api_base_url = "https://connect.squareup.com"
    
    return {
        "access_token": os.getenv("SQUARE_ACCESS_TOKEN", ""),
        "location_id": os.getenv("SQUARE_LOCATION_ID", ""),
        "api_base_url": os.getenv("SQUARE_API_BASE_URL", default_api_base_url),
        "environment": environment,
    }


def sanitize_for_logging(data):
    """Remove sensitive data before logging"""
    if not isinstance(data, dict):
        return data
    
    sanitized = data.copy()
    sensitive_keys = ["source_id", "token", "sourceId", "card", "cvv", "verification_token", "card_number"]
    
    for key in sensitive_keys:
        if key in sanitized:
            value = sanitized[key]
            if isinstance(value, str) and len(value) > 0:
                sanitized[key] = value[:TOKEN_PREVIEW_LENGTH] + "..." if len(value) > TOKEN_PREVIEW_LENGTH else "[REDACTED]"
            else:
                sanitized[key] = "[REDACTED]"
    
    return sanitized


def validate_payment_request(request_data):
    """Validate payment request data"""
    errors = []
    
    # Validate amount
    amount = request_data.get("amount")
    if amount is None:
        errors.append("amount is required")
    elif not isinstance(amount, (int, float, str)):
        errors.append("amount must be a number")
    else:
        try:
            amount_decimal = Decimal(str(amount))
            if amount_decimal <= 0:
                errors.append("amount must be greater than zero")
            elif amount_decimal < MIN_AMOUNT:
                errors.append(f"amount must be at least ${MIN_AMOUNT}")
            elif amount_decimal > MAX_AMOUNT:
                errors.append(f"amount exceeds maximum of ${MAX_AMOUNT}")
        except (InvalidOperation, ValueError):
            errors.append("amount must be a valid number")
    
    # Validate currency
    currency = request_data.get("currency", "USD")
    if not isinstance(currency, str) or len(currency) != 3 or not currency.isalpha():
        errors.append("currency must be a 3-letter ISO code (e.g., USD)")
    
    # Validate source_id
    source_id = request_data.get("source_id") or request_data.get("sourceId") or request_data.get("token")
    if not source_id:
        errors.append("source_id is required")
    elif not isinstance(source_id, str) or len(source_id) < 10:
        errors.append("source_id must be a valid payment token")
    
    return errors


def create_error_response(status_code, error_message, error_code=None, cors_headers=None):
    """Create standardized error response"""
    if cors_headers is None:
        cors_headers = {
            "Content-Type": "application/json",
            "Access-Control-Allow-Origin": "*",
            "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
            "Access-Control-Allow-Headers": "Content-Type, Authorization",
        }
    
    response_body = {
        "success": False,
        "error": {
            "message": error_message,
            "code": error_code or f"ERR_{status_code}",
        }
    }
    
    return {
        "statusCode": status_code,
        "headers": cors_headers,
        "body": json.dumps(response_body)
    }


def handle_process_payment(request_data, cors_headers):
    """Process Square payment"""
    try:
        print("[Square] Starting payment processing...")
        sanitized_data = sanitize_for_logging(request_data)
        print(f"[Square] Request data keys: {list(sanitized_data.keys()) if isinstance(sanitized_data, dict) else 'not a dict'}")
        print(f"[Square] Request data: {json.dumps(sanitized_data)[:LOG_TRUNCATE_LENGTH]}")
        
        # Get configuration at runtime
        config = get_square_config()
        square_access_token = config["access_token"]
        square_location_id = config["location_id"]
        square_api_base_url = config["api_base_url"]
        square_environment = config["environment"]
        
        print(f"[Square] Config loaded - Environment: {square_environment}, API URL: {square_api_base_url}")
        
        # Validate credentials
        if not square_access_token:
            return create_error_response(
                500,
                "SQUARE_ACCESS_TOKEN environment variable not set",
                "MISSING_CREDENTIALS",
                cors_headers
            )
        
        if not square_location_id:
            return create_error_response(
                500,
                "SQUARE_LOCATION_ID environment variable not set",
                "MISSING_CREDENTIALS",
                cors_headers
            )
        
        # Validate request data
        validation_errors = validate_payment_request(request_data)
        if validation_errors:
            return create_error_response(
                400,
                "; ".join(validation_errors),
                "VALIDATION_ERROR",
                cors_headers
            )
        
        # Extract and normalize request data
        source_id = request_data.get("source_id") or request_data.get("sourceId") or request_data.get("token")
        amount = request_data.get("amount")
        currency = request_data.get("currency", "USD").upper()
        idempotency_key = request_data.get("idempotency_key") or request_data.get("idempotencyKey") or request_data.get("orderId")
        
        print(f"[Square] Parsed request - source_id: {'present' if source_id else 'missing'}, amount: {amount}, currency: {currency}, idempotency_key: {'present' if idempotency_key else 'missing'}")
        
        # Generate idempotency key if not provided
        if not idempotency_key:
            idempotency_key = f"{int(time.time() * 1000)}-{''.join(random.choices('abcdefghijklmnopqrstuvwxyz0123456789', k=9))}"
            print(f"[Square] Generated idempotency_key: {idempotency_key}")
        
        # Convert amount to cents using Decimal for precision
        try:
            amount_decimal = Decimal(str(amount))
            amount_cents = int(amount_decimal * 100)
            print(f"[Square] Amount conversion: ${amount} → {amount_cents} cents")
        except (InvalidOperation, ValueError) as e:
            return create_error_response(
                400,
                f"Invalid amount format: {str(e)}",
                "INVALID_AMOUNT",
                cors_headers
            )
        
        if not REQUESTS_AVAILABLE or not requests:
            print("[Square API] ERROR: requests library not available")
            return create_error_response(
                500,
                "Server configuration error: requests library not available. Please ensure requirements.txt includes 'requests>=2.31.0'",
                "MISSING_DEPENDENCY",
                cors_headers
            )
        
        # Prepare Square API request
        square_payload = {
            "source_id": source_id,
            "idempotency_key": idempotency_key,
            "amount_money": {
                "amount": amount_cents,
                "currency": currency,
            },
            "location_id": square_location_id,
            "autocomplete": True,
        }
        
        # Add note if risk profile provided
        if request_data.get("risk_profile"):
            square_payload["note"] = f"Aave deposit - {request_data['risk_profile']} strategy"
        
        # Verify production endpoint
        api_url = f"{square_api_base_url}/v2/payments"
        is_production = "squareup.com" in square_api_base_url and "sandbox" not in square_api_base_url
        
        print(f"[Square] Environment: {square_environment}")
        print(f"[Square] API Base URL: {square_api_base_url}")
        print(f"[Square] Production Mode: {is_production}")
        print(f"[Square] Calling Square API: {api_url}")
        print(f"[Square] Amount: ${amount} ({amount_cents} cents)")
        print(f"[Square] Location ID: {square_location_id}")
        print(f"[Square] Access Token (first 10 chars): {square_access_token[:10] if square_access_token else 'NOT SET'}...")
        
        # Call Square Payments API
        response = requests.post(
            api_url,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {square_access_token}",
                "Square-Version": os.getenv("SQUARE_API_VERSION", "2024-10-16"),
            },
            json=square_payload,
            timeout=REQUEST_TIMEOUT,
        )
        
        print(f"[Square] Square API response status: {response.status_code}")
        
        # Handle response
        if not response.ok:
            error_data = {}
            try:
                if response.text:
                    error_data = response.json()
            except:
                error_data = {"detail": response.text or f"HTTP {response.status_code}"}
            
            errors = error_data.get("errors", [])
            if errors:
                error_detail = errors[0].get("detail", "Payment Failed.")
                error_code = errors[0].get("code", "UNKNOWN")
                print(f"[Square] Payment failed: {error_code} - {error_detail}")
                return create_error_response(
                    response.status_code,
                    error_detail,
                    f"SQUARE_{error_code}",
                    cors_headers
                )
            else:
                error_detail = error_data.get("detail", f"Square API error: {response.status_code}")
                return create_error_response(
                    response.status_code,
                    error_detail,
                    "SQUARE_API_ERROR",
                    cors_headers
                )
        
        # Parse successful response
        try:
            data = response.json()
        except Exception as e:
            print(f"[Square] Failed to parse JSON response: {e}")
            return create_error_response(
                500,
                "Invalid JSON response from Square API",
                "INVALID_RESPONSE",
                cors_headers
            )
        
        # Extract payment data from Square API response
        payment_data = data.get("payment", {})
        payment_id = payment_data.get("id")
        payment_status = payment_data.get("status")  # COMPLETED, APPROVED, FAILED, CANCELED
        
        print(f"[Square] Payment processed - ID: {payment_id}, Status: {payment_status}")
        
        # Return success response
        return {
            "statusCode": 200,
            "headers": cors_headers,
            "body": json.dumps({
                "success": True,
                "payment_id": payment_id,
                "status": payment_status,
                "order_id": payment_data.get("order_id"),
                "transaction_id": payment_id,
                "message": "Payment processed successfully",
                "amount_money": payment_data.get("amount_money", {}),
            })
        }
    
    except Exception as e:
        # Check if it's a timeout or connection error
        error_type = type(e).__name__
        error_module = type(e).__module__
        
        # Handle timeout errors
        if error_type == "Timeout" or (error_module == "requests.exceptions" and error_type == "Timeout"):
            print(f"[Square] Square API request timed out: {e}")
            return create_error_response(
                504,
                f"Square API request timed out: {str(e)}",
                "TIMEOUT_ERROR",
                cors_headers
            )
        # Handle connection errors
        elif error_type == "ConnectionError" or (error_module == "requests.exceptions" and error_type == "ConnectionError"):
            print(f"[Square] Cannot connect to Square API: {e}")
            return create_error_response(
                503,
                f"Cannot connect to Square API: {str(e)}",
                "CONNECTION_ERROR",
                cors_headers
            )
        else:
            # Generic exception handler
            print(f"[Square] Unexpected error: {e}")
            traceback.print_exc()
            return create_error_response(
                500,
                f"Internal server error: {str(e)}",
                "INTERNAL_ERROR",
                cors_headers
            )


def handler(event, context):
    """
    Vercel Python serverless function handler for /api/square/process-payment
    """
    # CORS headers
    cors_headers = {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type, Authorization",
    }
    
    # Handle None or invalid event
    if event is None:
        event = {}
    if not isinstance(event, dict):
        try:
            event = dict(event) if hasattr(event, '__dict__') else {}
        except:
            event = {}
    
    # Check for initialization errors
    if _INIT_ERROR:
        print(f"[Square API] Initialization error detected: {_INIT_ERROR}")
        return create_error_response(
            500,
            "Server configuration error",
            "INIT_ERROR",
            cors_headers
        )
    
    try:
        # Parse request method
        method = event.get("httpMethod") or event.get("method") or event.get("requestMethod") or "GET"
        
        # Handle OPTIONS (CORS preflight)
        if method == "OPTIONS":
            return {
                "statusCode": 200,
                "headers": cors_headers,
                "body": ""
            }
        
        # Only allow POST
        if method != "POST":
            return create_error_response(
                405,
                "Method not allowed. Use POST for payment processing.",
                "METHOD_NOT_ALLOWED",
                cors_headers
            )
        
        # Parse body
        body = event.get("body", "") or ""
        request_data = {}
        if body:
            try:
                if isinstance(body, str):
                    request_data = json.loads(body)
                else:
                    request_data = body
            except (json.JSONDecodeError, ValueError) as e:
                print(f"[Square API] Body parse error: {e}")
                return create_error_response(
                    400,
                    f"Invalid JSON in request body: {str(e)}",
                    "INVALID_JSON",
                    cors_headers
                )
        
        # Process payment
        return handle_process_payment(request_data, cors_headers)
    
    except Exception as e:
        error_type = type(e).__name__
        error_msg = str(e)
        
        print(f"[Square API] Handler error: {error_type}: {error_msg}")
        traceback.print_exc()
        
        return create_error_response(
            500,
            f"Internal server error: {error_msg}",
            "HANDLER_ERROR",
            cors_headers
        )

